File: android/src/main/kotlin/com/itoken/team/wutspider/impl/http/SpiderHttpClientImpl.kt
package com.itoken.team.wutspider.impl.http

import com.itoken.team.wutspider.SpiderException
import com.itoken.team.wutspider.interfaces.ISpiderHttpClient
import com.itoken.team.wutspider.model.EngineOptions
import com.itoken.team.wutspider.model.SpiderPayloadType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.security.KeyStore
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManagerFactory
import javax.net.ssl.X509TrustManager


class SpiderHttpClientImpl(options: EngineOptions) : ISpiderHttpClient {

    companion object {

        val RedirectHttpCodes = setOf(301, 302)

    }

    private val client = OkHttpClient.Builder()
        .followRedirects(false)
        .also {
            if (options.cookie) {
                it.cookieJar(InMemoryCookieJar())
            }
            if (options.forceSSL) {
                it.ignoreSSL()
            }
        }.build()

    private fun Request.Builder.applySpiderDefaults(): Request.Builder {
        return addHeader(
            "User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0"
        )
    }

    private fun makeRequest(
        request: Request,
        success: Int,
        allowRedirect: Boolean,
    ): ISpiderHttpClient.HttpClientResponse {
        return runCatching {
            var resp = client.newCall(request).execute()
            while (allowRedirect) {
                val newRequest = RedirectHandling.followUpRequest(resp)
                if (newRequest != null) {
                    resp = client.newCall(request).execute()
                }
            }
            resp
        }.map { resp ->
            resp.use { response ->
                if (response.code != success) {
                    throw SpiderException("Unexpected status code while requesting ${request.url}. actual: ${response.code}, expected: $success")
                }
                val content = response.body?.string() ?: ""
                val headerNames = response.headers.names()
                val headerValues = headerNames.map {
                    response.headers.values(it).joinToString(";")
                }
                val zipped = headerNames.zip(headerValues).toTypedArray()
                val headers = mapOf(*zipped)
                ISpiderHttpClient.HttpClientResponse(
                    response.code,
                    headers,
                    content
                )
            }
        }.getOrElse {
            throw SpiderException("Failed to request url ${request.url}", it)
        }
    }

    override fun get(
        url: String,
        headers: Map<String, String>,
        success: Int,
        autoRedirect: Boolean
    ): ISpiderHttpClient.HttpClientResponse {
        val request = Request.Builder()
            .url(url)
            .method("GET", null)
            .applySpiderDefaults()
            .also {
                headers.forEach { (key, value) -> it.addHeader(key, value) }
            }
            .build()
        return makeRequest(request, success, autoRedirect)
    }

    override fun post(
        url: String,
        headers: Map<String, String>,
        payload: String,
        payloadType: SpiderPayloadType,
        success: Int,
        autoRedirect: Boolean
    ): ISpiderHttpClient.HttpClientResponse {
        val mediaType = when (payloadType) {
            SpiderPayloadType.JSON -> "application/json".toMediaTypeOrNull()
            SpiderPayloadType.FORM -> "application/x-www-form-urlencoded".toMediaTypeOrNull()
            else -> "application/text".toMediaTypeOrNull()
        }
        val request = Request.Builder()
            .url(url)
            .method("POST", payload.toRequestBody(mediaType))
            .applySpiderDefaults()
            .also {
                headers.forEach { (key, value) -> it.addHeader(key, value) }
            }
            .build()
        return makeRequest(request, success, autoRedirect)
    }

    private fun OkHttpClient.Builder.ignoreSSL() {
        val sslContext = SSLContext.getInstance("SSL")
        val trustManager = object : X509TrustManager {
            override fun checkClientTrusted(
                chain: Array<out X509Certificate>?,
                authType: String?
            ) {
            }

            override fun checkServerTrusted(
                chain: Array<out X509Certificate>?,
                authType: String?
            ) {
            }

            override fun getAcceptedIssuers() = emptyArray<X509Certificate>()
        }
        sslContext.init(
            null, arrayOf(trustManager), SecureRandom()
        )
        hostnameVerifier { _, _ -> true }
        sslSocketFactory(sslContext.socketFactory, platformTrustManager())
    }

    private fun platformTrustManager(): X509TrustManager {
        val factory = TrustManagerFactory.getInstance(
            TrustManagerFactory.getDefaultAlgorithm()
        )
        factory.init(null as KeyStore?)
        val trustManagers = factory.trustManagers!!
        check(trustManagers.size == 1 && trustManagers[0] is X509TrustManager) {
            "Unexpected default trust managers: ${trustManagers.contentToString()}"
        }
        return trustManagers[0] as X509TrustManager
    }

}

File: ios/src/Implement/Http/SpiderHttpClientImpl.swift
//
//  Untitled.swift
//  SpiderEngine
//
//  Created by 李卓强 on 2025/3/25.
//

import Foundation

class SpiderHttpClientImpl : SpiderHttpClient {
    private let session: URLSession
    private let userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36 Edg/122.0.0.0"
    
    init(engineOptions: EngineOptions) {
        let configuration = URLSessionConfiguration.default
        configuration.httpShouldSetCookies = engineOptions.cookie
        configuration.httpCookieAcceptPolicy = engineOptions.cookie ? .always : .never
        configuration.httpShouldUsePipelining = true
        
        self.session = URLSession(configuration: configuration)
    }
    
    func getAsync(url: String, headers: [String: String], success: Int = 200) async throws -> (content: String, headers: [String: String]) {
        guard let url = URL(string: url) else {
            throw HttpError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        headers.forEach { request.setValue($0.value, forHTTPHeaderField: $0.key) }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw HttpError.invalidResponse
        }
        
        guard httpResponse.statusCode == success else {
            throw HttpError.requestFailed(statusCode: httpResponse.statusCode, expected: success)
        }
        
        let content = String(data: data, encoding: .utf8) ?? ""
        let responseHeaders : [String: String] = Dictionary(uniqueKeysWithValues:
            httpResponse.allHeaderFields.compactMap { key, value in
                guard let keyString = key as? String,
                      let valueString = value as? String else { return nil }
                return (keyString, valueString)
            }
        )
        
        return (content, responseHeaders)
    }
    
    func postAsync(url: String, headers: [String: String], payload: String, type: SpiderPayloadType, success: Int = 200) async throws -> (content: String, headers: [String: String]) {
        guard let url = URL(string: url) else {
            throw HttpError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        
        let contentType: String
        switch type {
        case .Json:
            contentType = "application/json"
        case .Form:
            contentType = "application/x-www-form-urlencoded"
        default:
            contentType = "application/text"
        }
        
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        request.httpBody = payload.data(using: .utf8)
        headers.forEach { request.setValue($0.value, forHTTPHeaderField: $0.key) }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw HttpError.invalidResponse
        }
        
        guard httpResponse.statusCode == success else {
            throw HttpError.requestFailed(statusCode: httpResponse.statusCode, expected: success)
        }
        
        let content = String(data: data, encoding: .utf8) ?? ""
        let responseHeaders : [String: String] = Dictionary(uniqueKeysWithValues:
            httpResponse.allHeaderFields.compactMap { key, value in
                guard let keyString = key as? String,
                      let valueString = value as? String else { return nil }
                return (keyString, valueString)
            }
        )
        
        return (content, responseHeaders)
    }
}
